pkg/tso/keyspace_group_manager.go

```diff
@@ -439,22 +481,19 @@ func (kgm *KeyspaceGroupManager) isAssignedToMe(group *endpoint.KeyspaceGroup) b
 // updateKeyspaceGroup applies the given keyspace group. If the keyspace group is just assigned to
 // this host/pod, it will join the primary election.
 func (kgm *KeyspaceGroupManager) updateKeyspaceGroup(group *endpoint.KeyspaceGroup) {
-	if group.ID >= uint32(len(kgm.ams)) {
-		log.Warn("keyspace group ID is out of range, ignore it",
-			zap.Uint32("keyspace-group-id", group.ID), zap.Int("max-keyspace-group-id", len(kgm.ams)-1))
+	if err := kgm.checkKeySpaceGroupID(group.ID); err != nil {
+		log.Warn("keyspace group ID is invalid, ignore it", zap.Error(err))
 		return
 	}
 
 	assignedToMe := kgm.isAssignedToMe(group)
 	if assignedToMe {
-		if kgm.ams[group.ID].Load() != nil {
+		if kgm.ams[group.ID] != nil {
```

Comment:
No need. 

There are three writers -- initialization (Load the initial keyspace group assignment from storage), KeyspaceGroupsMetaWatchLoop (the code shown above is in this single goroutine), and shutdown. They are serialized, i.e., there are no concurrent writers. Specifically, we start KeyspaceGroupsMetaWatchLoop after static initialization, and the system exits from KeyspaceGroupsMetaWatchLoop before write access in shutdown.

In summary, KeyspaceGroupsMetaWatchLoop can read without lock, but it must require lock for write so that keyspace group manager can handle concurrent reads when processing requests.

Reply:
But I suggest adding one to prevent the potential problem, it doesn't cost too much I think.

Reply:
It has cost. If I add a lock it needs to be a write lock and adding here will increase the length of the critical section.

Reply:
The lock isn't re-entrant. If I'm right, with sync.RWMutex, I even can't require a read lock then upgrade it the write lock.

Reply:
good suggestion, actually.

changed to (kgm.getAllocatorManager requires the read lock)
```
		if oldAM, oldGroup := kgm.getAllocatorManager(group.ID); oldAM != nil {
			log.Info("keyspace group already initialized, so update meta only",
				zap.Uint32("keyspace-group-id", group.ID))
			kgm.updateKeyspaceGroupMembership(oldGroup, group)
			return
		}
```